### ISSUE TITLE
How to find automation_id of element whose control type is text in java based application?

### ISSUE THREAD

**Issue by indrajeet7** [pywinauto/pywinauto#744]
I have java based GUI application whose automation_ids(JavaFX) are changing continuously. So, I cant use print_control_identifiers(). So, to click on button how to find its automation_id or mouse coordinate(rectangle function is also not working)?

**Comment by vasily-v-ryabov** [comment]
Can't believe that "everything is not working". What is exactly tried? can you provide at least pseudo-code and error output? What `print_control_identifiers()` or `dump_tree()` is printing?

**Comment by indrajeet7** [comment]
print_control_identifiers() and dump_tree() are working. But, it takes lot of time(more than 2 hours) to print.

**Comment by indrajeet7** [comment]
How can I reduce this time? Also, I have java application after clicking on button it opens pop up window. 
How can I get control_properties for this window only?

**Comment by vasily-v-ryabov** [comment]
Yes, we know about performance problems with rich apps. It's possible to use `dump_tree(depth=2)` for immediate children only or use methods `descendants()` and `children()` to walk the hierarchy in a loop.

**Comment by indrajeet7** [comment]
Thanks for your reply. Can you explain me how to use descendants() and children() to walk the hierarchy in a loop with some demo example?

**Comment by indrajeet7** [comment]
How to store output of descendants() and children() into .txt file?

**Comment by vasily-v-ryabov** [comment]
`.descendants()` and `.children()` can be called the same way as `.dump_tree()`. They return a list of wrappers. Each wrapper has methods like `.window_text()`, `.automation_id()`, `.control_type()` etc.

It's easy to find how to write any string to file using Python. We don't reply on general Python questions. Let's ask someone other or just google. Thanks for understanding.

**Comment by indrajeet7** [comment]
With descendants() automation id for F0/D0 button is JavaFX569 and with print_control_identifiers() automation id for F0/D0 button is JavaFX574?

I am able to click on F0/D0 button with automation_id obtained with print_control_identifiers()  only not with descendants().
Why I got different automation id with print_control_identifiers() and with descendants()?

Below I pasted some of output with print_control_identifiers() and with descendants() - 

Output with descendants() is as below - 

{'class_name': '', 'friendly_class_name': 'Button', 'texts': [''], 'control_id': None, 'rectangle': <RECT L736, T144, R776, B174>, 'is_visible': True, 'is_enabled': True, 'control_count': 1, 'is_keyboard_focusable': True, 'has_keyboard_focus': False, 'automation_id': 'JavaFX565'}
{'class_name': '', 'friendly_class_name': 'Button', 'texts': ['F0/D0'], 'control_id': None, 'rectangle': <RECT L0, T169, R50, B193>, 'is_visible': True, 'is_enabled': True, 'control_count': 1, 'is_keyboard_focusable': True, 'has_keyboard_focus': False, 'automation_id': 'JavaFX569'}
{'class_name': '', 'friendly_class_name': 'Button', 'texts': ['F1/D0'], 'control_id': None, 'rectangle': <RECT L51, T169, R100, B193>, 'is_visible': True, 'is_enabled': True, 'control_count': 1, 'is_keyboard_focusable': True, 'has_keyboard_focus': False, 'automation_id': 'JavaFX570'}
{'class_name': '', 'friendly_class_name': 'Button', 'texts': ['F2/D0'], 'control_id': None, 'rectangle': <RECT L101, T169, R150, B193>, 'is_visible': True, 'is_enabled': True, 'control_count': 1, 'is_keyboard_focusable': True, 'has_keyboard_focus': False, 'automation_id': 'JavaFX571'}

Output with print_control_identifiers() is as below - 
   |    | 
   |    | Image - ''    (L-31245, T-31842, R-31226, B-31822)
   |    | ['Image9', 'Block View(FTL mode)Image2']
   |    | child_window(auto_id="JavaFX570", control_type="Image")
   | 
   | Button - 'F0/D0'    (L-31990, T-31822, R-31940, B-31798)
   | ['F0/D0', 'Button43', 'F0/D0Button', 'F0/D00', 'F0/D01']
   | child_window(title="F0/D0", auto_id="JavaFX574", control_type="Button")
   |    | 
   |    | Static - 'F0/D0'    (L-31980, T-31817, R-31951, B-31803)
   |    | ['Static92', 'F0/D02', 'F0/D0Static', 'F0/D0Static0', 'F0/D0Static1']
   |    | child_window(title="F0/D0", auto_id="JavaFX575", control_type="Text")
   | 
   | Button - 'F1/D0'    (L-31940, T-31822, R-31891, B-31798)
   | ['F1/D0', 'F1/D0Button', 'Button44', 'F1/D00', 'F1/D01']
   | child_window(title="F1/D0", auto_id="JavaFX576", control_type="Button")
   |    | 
   |    | Static - 'F1/D0'    (L-31930, T-31817, R-31901, B-31803)
   |    | ['F1/D02', 'Static93', 'F1/D0Static']
   |    | child_window(title="F1/D0", auto_id="JavaFX577", control_type="Text")
   | 
   | Button - 'F2/D0'    (L-31890, T-31822, R-31841, B-31798)
   | ['F2/D0', 'Button45', 'F2/D0Button', 'F2/D00', 'F2/D01']
   | child_window(title="F2/D0", auto_id="JavaFX578", control_type="Button")
   |    | 
   |    | Static - 'F2/D0'    (L-31880, T-31817, R-31851, B-31803)
   |    | ['Static94', 'F2/D02', 'F2/D0Static']
   |    | child_window(title="F2/D0", auto_id="JavaFX579", control_type="Text")

**Comment by airelil** [comment]
It would be interesting to know what  are values of automation_id and control coordinates reported by Inspector.exe utility running in "UIA mode"

**Comment by vasily-v-ryabov** [comment]
Guys on Reddit recommend TestFX tool for JavaFX apps:
https://www.reddit.com/r/JavaFX/comments/7naiyv/is_it_possible_to_set_the_automationid_on_a/

It seems automation_id is not guaranteed to be unique for JavaFX. Some day we may try to investigate technical possibility to automate Java apps better way. But not right now.

**Comment by indrajeet7** [comment]
Inspect.exe generates same automation_id same as that of print_control_identifiers()

**Comment by karges612** [comment]
@indrajeet7  its a delayed comment , but wana understand how did you automate the JFX application ? was pywinauto usefull or did you choose some other solution?
